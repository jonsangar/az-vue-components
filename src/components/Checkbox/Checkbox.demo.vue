<template lang="pug">
  div
    Demo(title="Uso Básico")
      az-checkbox(v-model="checked") Checkbox

      p.margin-left.margin-top-half
        span.color-gray3 valor:&nbsp;
        span.color-brand {{ checked }}

    Demo(title="Desactivado")
      az-checkbox(v-model="disable1" disabled) Checkbox
      az-checkbox(v-model="disable2" disabled) Checkbox

    Demo(title="Grupos")
      az-checkbox-group
        az-checkbox(
          v-for="(item, index) in sports"
          v-model="groupResult"
          :key="index"
          :value="item")
          | {{ item }}

      p.margin-left.margin-top-half
        span.color-gray3 valor:&nbsp;
        span.color-brand {{ groupResult }}

    Demo(title="Máximo de checkbox seleccionados")
      az-checkbox-group(max-options="2")
        az-checkbox(
          v-for="(item, index) in sports"
          v-model="groupMax"
          :key="index"
          :value="item")
          | {{ item }}

      p.margin-left.margin-top-half
        span.color-gray3 valor:&nbsp;
        span.color-brand {{ groupMax }}

    Demo.demo--cell.no-margin-bottom(title="Vista modo celda")
      az-checkbox-group(cell)
        az-checkbox(
          v-for="(item, index) in sports"
          v-model="groupCell"
          :key="index"
          :value="item")
          | {{ item }}

      p.margin-left.margin-top-half
        span.color-gray3 valor:&nbsp;
        span.color-brand {{ groupCell }}
</template>

<script>
export default {
  data() {
    return {
      checked: true,
      disable1: false,
      disable2: true,
      sports: ['swimming', 'cycling', 'running'],
      groupResult: ['cycling'],
      groupMax: [],
      groupCell: [],
    };
  },
};
</script>

<style lang="scss"></style>
