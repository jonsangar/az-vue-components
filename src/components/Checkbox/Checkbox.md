---
device: true
---

# Fieldset

### Instalación

```javascript
import { Checkbox } from 'didor';

Vue.use(Checkbox);
```

---

### Uso

#### Uso básico

``` html
<az-checkbox v-model="checked">Checkbox</az-checkbox>
```

#### Desactivado

``` html
<az-checkbox v-model="checked" disabled>Checkbox</az-checkbox>
<az-checkbox v-model="checked" disabled>Checkbox</az-checkbox>
```

#### Grupos

Cuando varios checkbox están dentro de un grupo, el valor del grupo es un array con los `value` de cada checkbox.

``` html
<az-checkbox-group>
  <az-checkbox v-model="result" value="swimming">Natación</az-checkbox>
  <az-checkbox v-model="result" value="cycling">Ciclismo</az-checkbox>
  <az-checkbox v-model="result" value="running">Carrera</az-checkbox>
</az-checkbox-group>
```

Puedes generar un listado de checkbox con un array de valores:

``` html
<az-checkbox-group>
  <az-checkbox
    v-for="(item, index) in sports"
    v-model="result"
    :key="index"
    :value="item">
      {{ item }
  </az-checkbox>
</az-checkbox-group>
```

``` js
export default {
  data() {
    return {
      sports: ['swimming', 'cycling', 'running'],
      result: ['cycling'],
    };
  },
};
```

#### Máximo número de checkbox seleccionados

Cuando varios checkbox están dentro de un grupo, el valor del grupo es un array con los `value` de cada checkbox.

``` html
<az-checkbox-group max="2">
  <az-checkbox v-model="result" value="swimming">Natación</az-checkbox>
  <az-checkbox v-model="result" value="cycling">Ciclismo</az-checkbox>
  <az-checkbox v-model="result" value="running">Carrera</az-checkbox>
</az-checkbox-group>
```

#### Modo celda

Modifica su visualización para verse como dentro de una celda

``` html
<az-checkbox-group cell>
  <az-checkbox v-model="result" value="swimming">Natación</az-checkbox>
  <az-checkbox v-model="result" value="cycling">Ciclismo</az-checkbox>
  <az-checkbox v-model="result" value="running">Carrera</az-checkbox>
</az-checkbox-group>
```

---

### API

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| title | Título del grupo | `String` | -- |

### Slots

| Nombre | Descripción                                             |
| ------ | ------------------------------------------------------- |
| title | Título del grupo |
