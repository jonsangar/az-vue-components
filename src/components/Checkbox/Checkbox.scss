$checkboxSize: size(0.8) !default;
$checkboxSizeCell: size(1.05) !default;
$checkboxBackground: #ffffff !default;
$checkboxCheck: color(brand) !default;
$checkboxBorderColor: color(gray5) !default;
$checkboxBorder: color(gray7) !default;
$checkboxBorderWidth: 2px !default;
$checkboxBorderRadius: 2px !default;
$checkboxMarkColor: #ffffff !default;

.az-checkbox {
  $rootClass: &;

  display: flex;
  align-items: center;
  overflow: hidden;
  cursor: pointer;
  user-select: none;


  &__input {
    position: relative;
  }

  &__base {
    width: $checkboxSize;
    height: $checkboxSize;
    margin-right: half();
    cursor: pointer;
    background: $checkboxBackground;
    border: $checkboxBorderWidth solid $checkboxBorderColor;
    border-radius: $checkboxBorderRadius;
    transition: border 250ms cubic-bezier(0.4, 0, 0.23, 1);
  }

  &__mark {
    position: absolute;
    top: 0.6em;
    left: 0.2em;
    z-index: 1;
    width: 0.5em;
    height: 0.75em;
    content: '';
    border-top: 0.2em solid $checkboxMarkColor;
    border-right: 0.2em solid $checkboxMarkColor;
    opacity: 0;
    transition: all 250ms cubic-bezier(0.4, 0, 0.23, 1);
    transition-delay: 200ms;
    transform: scaleX(-1) rotate(135deg);
    transform-origin: left top;
  }

  &__cell {
    position: relative;
    display: none;
    width: $checkboxSizeCell;
    height: $checkboxSizeCell;
    border: 2px solid $checkboxBorder;
    border-radius: 50%;
    transition: transform 0.14s ease;
    transform-origin: 50% 50%;
    transform-style: preserve-3d;

    svg {
      position: absolute;
      top: -2px;
      right: -2px;
      bottom: -2px;
      left: -2px;
      z-index: 1;
      display: block;
      width: $checkboxSizeCell;
      height: $checkboxSizeCell;
      fill: none;
      stroke: $checkboxBorderColor;
      stroke-dasharray: 0 162.6 133 (162.6 - 133);
      stroke-dashoffset: 162.6 - 38;
      stroke-linecap: round;
      stroke-linejoin: round;
      stroke-width: 3;
      transition: all 0.4s ease 0s;
    }


  }

  &__group__header{
    background-color: color(gray7) !important;
    color: black !important;
  }

  &__label {
    line-height: simple();
    color: color(content);
  }

  //
  // Checked State
  //
  &--checked {
    #{$rootClass}__base {
      border: (calc($checkboxSize / 2) + 0.001) solid color(brand);
    }

    #{$rootClass}__mark {
      opacity: 1; // Se descuadra en las cabeceras de cell group de InviteUser
      // opacity: 0; // TODO: Arreglarlo para que est√© a 1 sin descuadrar
    }

    #{$rootClass}__cell {
      svg {
        stroke: $checkboxCheck;
        stroke-dasharray: 0 162.6 28 (162.6 - 28);
        stroke-dashoffset: 162.6;
        transition: all 0.4s ease 0.2s;
      }
    }
  }

  //
  // Disabled State
  //
  &--disabled {
    #{$rootClass}__label {
      color: color(gray4);
    }

    #{$rootClass}__input {
      opacity: 0.5;
    }

    #{$rootClass}__label {
      opacity: 0.7;
    }
  }

  @include hover {
    .az-checkbox {
      &__base {
        background: rgba(255, 255, 255, 0.1);
      }
    }
  }
}
