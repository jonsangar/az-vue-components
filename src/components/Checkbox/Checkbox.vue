<template lang="pug">
  .az-checkbox(:class="classObject" @click="onInput")
    .az-checkbox__cell
      svg(viewBox="0 0 44 44")
        path(d="M14,24 L21,31 L39.7428882,11.5937758 C35.2809627,6.53125861 30.0333333,4 24,4 C12.95,4 4,12.95 4,24 C4,35.05 12.95,44 24,44 C35.05,44 44,35.05 44,24 C44,19.3 42.5809627,15.1645919 39.7428882,11.5937758" transform="translate(-2.000000, -2.000000)")
    .az-checkbox__input
      .az-checkbox__base
      .az-checkbox__mark
    .az-checkbox__label
      slot
</template>

<script>
export default {
  props: {
    value: {
      type: [String, Number, Boolean],
    },
    checked: {
      type: [Boolean, Array],
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    disabledStyle: {
      type: Boolean,
      default: false,
    },
  },
  model: {
    prop: 'checked',
    event: 'input',
  },
  computed: {
    /**
     * Devuelve un listado con las clases que dependen de la configuración
     */
    classObject() {
      return [
        this.disabledStyle ? 'az-checkbox--disabled' : '',
        this.disabled ? 'az-checkbox--disabled' : '',
        this.isChecked ? 'az-checkbox--checked' : '',
      ];
    },

    isGroup() {
      return Array.isArray(this.checked);
    },

    isChecked() {
      if (this.isGroup) return this.checked.includes(this.value);
      return this.checked;
    },
  },
  methods: {
    /**
     * Evento lanzado cuando se modifica el valor del checkbox
     */
    onInput() {
      if (this.disabled) return;

      if (!this.isGroup || this.disabledStyle) {
        this.$emit('input', !this.checked);
        return;
      }

      // Creo un nuevo array para no hacer una mutación en la propiedad checked
      const checkedList = [].concat(this.checked);

      if (checkedList.includes(this.value)) {
        checkedList.splice(checkedList.indexOf(this.value), 1);
      } else {
        // Compruebo si existe un número máximo de opciones definidas en el grupo
        // y si ya se ha superado
        const maxOptions = this.$parent.$options.propsData.maxOptions;
        const options = this.checked.length;

        if (maxOptions && maxOptions <= options) return;

        checkedList.push(this.value);
      }

      this.$emit('input', checkedList);
    },
  },
};
</script>

<style src="./Checkbox.scss" lang="scss"></style>
