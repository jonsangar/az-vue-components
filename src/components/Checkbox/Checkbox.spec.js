import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import Checkbox from './Checkbox.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('Checkbox.vue', () => {
  const build = () => {
    const wrapper = shallowMount(Checkbox, {
      localVue,
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente Checkbox con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Comprueba que emite evento "click" al ser pulsado', async () => {
    const { wrapper } = build();
    // Al pulsar sobre el elemento lanza el evento click
    wrapper.trigger('click');
    // Esperamos para que emits sea manejado
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    // Comprobamos que se ha emitido el evento click
    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input.length).toBe(1);
    expect(wrapper.emitted().input[0]).toEqual([true])
  });

  it('Comprueba que no emite evento "click" al ser pulsado, cuando está deshabilitado', async () => {
    const { wrapper } = build();
    // Si actualizamos la propiedad disabled, no emite evento click
    await wrapper.setProps({
      disabled: true,
    });
    // Al pulsar sobre el elemento lanza el evento click
    wrapper.trigger('click');
    // Esperamos para que emits sea manejado
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    // Comprobamos que se ha emitido el evento click
    expect(wrapper.emitted().input).toBe(undefined);
  });

  // it('Comprueba que tiene el titulo', async () => {
  //   const { wrapper } = build();
  //   // Si actualizamos la propiedad title, añade el título
  //   await wrapper.setProps({
  //     title: 'myTitle',
  //   });
  //   // Esperamos para que se actualice la prop
  //   await wrapper.vm.$nextTick()
  //   // Comprobamos que se ha activado la prop
  //   expect(wrapper.vm.title).toStrictEqual('myTitle');
  //   // Comprobamos que se ha añadido el titulo con estilos por defecto
  //   expect(wrapper.find('.az-cell-group__title').isVisible()).toBe(true);
  //   expect(wrapper.find('.az-cell-group__title').text()).toBe('myTitle');
  // });
})
