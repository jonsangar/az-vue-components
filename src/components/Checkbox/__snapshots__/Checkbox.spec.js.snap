// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox.vue Comprueba que emite evento "click" al ser pulsado 1`] = `
<div class="az-checkbox">
  <div class="az-checkbox__cell"><svg viewBox="0 0 44 44">
      <path d="M14,24 L21,31 L39.7428882,11.5937758 C35.2809627,6.53125861 30.0333333,4 24,4 C12.95,4 4,12.95 4,24 C4,35.05 12.95,44 24,44 C35.05,44 44,35.05 44,24 C44,19.3 42.5809627,15.1645919 39.7428882,11.5937758" transform="translate(-2.000000, -2.000000)"></path>
    </svg></div>
  <div class="az-checkbox__input">
    <div class="az-checkbox__base"></div>
    <div class="az-checkbox__mark"></div>
  </div>
  <div class="az-checkbox__label"></div>
</div>
`;

exports[`Checkbox.vue Comprueba que no emite evento "click" al ser pulsado, cuando est√° deshabilitado 1`] = `
<div class="az-checkbox az-checkbox--disabled">
  <div class="az-checkbox__cell"><svg viewBox="0 0 44 44">
      <path d="M14,24 L21,31 L39.7428882,11.5937758 C35.2809627,6.53125861 30.0333333,4 24,4 C12.95,4 4,12.95 4,24 C4,35.05 12.95,44 24,44 C35.05,44 44,35.05 44,24 C44,19.3 42.5809627,15.1645919 39.7428882,11.5937758" transform="translate(-2.000000, -2.000000)"></path>
    </svg></div>
  <div class="az-checkbox__input">
    <div class="az-checkbox__base"></div>
    <div class="az-checkbox__mark"></div>
  </div>
  <div class="az-checkbox__label"></div>
</div>
`;

exports[`Checkbox.vue Comprueba renderizado del componente Checkbox con snapshot 1`] = `
<div class="az-checkbox">
  <div class="az-checkbox__cell"><svg viewBox="0 0 44 44">
      <path d="M14,24 L21,31 L39.7428882,11.5937758 C35.2809627,6.53125861 30.0333333,4 24,4 C12.95,4 4,12.95 4,24 C4,35.05 12.95,44 24,44 C35.05,44 44,35.05 44,24 C44,19.3 42.5809627,15.1645919 39.7428882,11.5937758" transform="translate(-2.000000, -2.000000)"></path>
    </svg></div>
  <div class="az-checkbox__input">
    <div class="az-checkbox__base"></div>
    <div class="az-checkbox__mark"></div>
  </div>
  <div class="az-checkbox__label"></div>
</div>
`;
